R/construct_api_requests.R

```diff
@@ -0,0 +1,296 @@
+    POST = TRUE
+  }
+  
+  time_periods <- c("last_modified", "datetime", "time")
+  if(any(time_periods %in% names(get_list))){
+    for(i in time_periods){
+      get_list[[i]] <- format_api_dates(get_list[[i]])
+      full_list[[i]] <- format_api_dates(full_list[[i]]) 
+    }
+  }
+  
+  baseURL <- explode_query(baseURL, POST = FALSE, get_list)
+  
+  if(all(!is.na(bbox))){
+    baseURL <- httr2::req_url_query(baseURL,
+                                    bbox = bbox,
+                                    .multi = "comma")      
+  }
+  
+  if(!all(is.na(properties))){
+    baseURL <- httr2::req_url_query(baseURL,
+                                    properties = properties,
+                                    .multi = "comma")    
+  }
+  
+  if(POST){  
+    baseURL <- baseURL |>
+      httr2::req_headers(`Content-Type` = "application/query-cql-json") 
+    
+    post_params <- list(
+      "params" = unname(post_params)
+    )
+    
+    x <- whisker::whisker.render(template_post, post_params)
+    baseURL <- httr2::req_body_raw(baseURL, x) 
+    
+  } else {
+    baseURL <- explode_query(baseURL, POST = FALSE, full_list)
+  }
+  
+  return(baseURL)
+}
+
+setup_api <- function(service){
+  
+  baseURL <- httr2::request("https://api.waterdata.usgs.gov/ogcapi/v0/collections") |> 
+    httr2::req_url_path_append(service, "items") |> 
+    basic_request() 
+  
+}
+
+format_api_dates <- function(datetime){
+  
+  if(!any(isTRUE(is.na(datetime)) | isTRUE(is.null(datetime)))){
+    if(length(datetime) == 1){
+      datetime <- format(datetime, format = "%Y-%m-%dT%H:%M:%SZ")
+    } else if (length(datetime) == 2) {
+      datetime <- as.POSIXct(datetime)
+      datetime <- paste0(vapply(datetime, FUN =  function(x) {
+        format(x, format = "%Y-%m-%dT%H:%M:%SZ")},
+        FUN.VALUE =  c(NA_character_)
+      ), collapse = "/")
+      datetime <- gsub("NA", "..", datetime)
+    } else {
+      stop("datetime should only include 1-2 values")
+    }
+  }
+  return(datetime)
+}
+
+explode_post <- function(ls){
+  
+  ls <- Filter(Negate(anyNA), ls)
+  params <- NULL
+  
+  if(max(lengths(ls)) > 1) {
+    
+    for(i in seq_along(ls)){
+      params[names(ls[i])] <- cql2_param(ls[i])
+    }
+    
+    if(length(params) > 1){
+      params[seq_along(1:(length(params)-1))] <- paste0(params[seq_along(1:(length(params)-1))], ",")
+    }
+  }
+  return(params)  
+}
+
+cql2_param <- function(parameter){
+  template_path <- system.file("templates/param.CQL2", package = "dataRetrieval")
+  template <- readChar(template_path, file.info(template_path)$size)
+  
+  parameters <- paste0(unlist(parameter), collapse = '", "')
+  parameters <- paste0('"', parameters, '"')
+  parameter_list <- list("property" = names(parameter),
+                         "parameter" = parameters)
+  return(whisker::whisker.render(template, parameter_list))
+} 
+
+#' Check OGC requests
+#' 
+#' @param endpoint Character, can be "daily", "timeseries-metadata"
```

Comment:
```suggestion
#' @param endpoint Character, can be "daily", "time-series-metadata", "monitoring-locations"
```

Might be worth genericizing to "can be any existing collection"?